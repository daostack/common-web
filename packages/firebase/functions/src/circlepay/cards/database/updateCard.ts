import { firestore } from 'firebase-admin';

import { ICardEntity } from '../types';
import { CardCollection } from './index';


/**
 * Updates the passed entity in the database. Reference
 * is kept from the ID of the entity, so if the ID is changed (please,
 * don't do this) new entity will be created
 *
 * @param card - The card entity to update
 *
 * @returns - The updated card entity
 */
export const updateCardInDatabase = async (card: ICardEntity): Promise<ICardEntity> => {
  const cardDoc = {
    ...card,

    updatedAt: firestore.Timestamp.now()
  };

  await CardCollection
    .doc(cardDoc.id)
    .update(cardDoc);

  return cardDoc;
}