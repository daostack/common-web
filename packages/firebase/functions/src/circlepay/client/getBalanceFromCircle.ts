import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { getCircleHeaders } from '../index';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';


interface ICircleAmount {
  amount: string;
  currency: string;
}

interface ICircleBalance {
  data: {
    available: ICircleAmount[];
    unsettled: ICircleAmount[];
  }
}

export const getBalanceFromCircle = async (): Promise<ICircleBalance> => {
  const headers = await getCircleHeaders();

  return externalRequestExecutor<ICircleBalance>(async () => {
    return (await axios.get<ICircleBalance>(`${circlePayApi}/businessAccount/balances`, headers)).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    message: 'Failed getting balance from circle'
  });
}