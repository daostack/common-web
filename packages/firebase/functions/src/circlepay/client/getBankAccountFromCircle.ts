import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';

import { ICircleGetBankAccountResponse } from '../cards/circleTypes';
import { getCircleHeaders } from '../index';

export const getBankAccountFromCircle = async (bankAccountId: string): Promise<ICircleGetBankAccountResponse> => {
  const headers = await getCircleHeaders();

  return  await externalRequestExecutor<ICircleGetBankAccountResponse>(async () => {
    return (await axios.get<ICircleGetBankAccountResponse>(`${circlePayApi}/banks/wires/${bankAccountId}`,
      headers
    )).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    userMessage: 'Cannot get the bank account'
  });
}