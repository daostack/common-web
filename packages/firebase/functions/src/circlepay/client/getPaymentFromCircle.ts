import axios from 'axios';

import { externalRequestExecutor } from '../../util';
import { circlePayApi } from '../../settings';
import { ErrorCodes } from '../../constants';

import { getCircleHeaders } from '../index';
import { ICirclePayment } from '../types';

/**
 * Gets the current state of the payment from Circle
 *
 * @param paymentId - the Circle payment ID (not the local one)
 */
export const getPaymentFromCircle = async (paymentId: string): Promise<ICirclePayment> => {
  const headers = await getCircleHeaders();

  return externalRequestExecutor<ICirclePayment>(async () => {
    return (await axios.get<ICirclePayment>(`${circlePayApi}/payments/${paymentId}`, headers)).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    message: `Circle call to GET payment with id ${paymentId} failed`
  });
};

