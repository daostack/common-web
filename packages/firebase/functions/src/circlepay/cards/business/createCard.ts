import axios from 'axios';
import * as yup from 'yup';

import { ErrorCodes } from '../../../constants';
import { circlePayApi } from '../../../settings';
import { validate } from '../../../util/validate';
import { externalRequestExecutor } from '../../../util';
import { billingDetailsValidationSchema } from '../../../util/schemas';

import { ICircleCreateCardPayload, ICircleCreateCardResponse } from '../../types';
import { getCircleHeaders } from '../../index';
import { ICardEntity } from '../types';
import { userDb } from '../../../users/database';
import { cardDb } from '../database';
import { createEvent } from '../../../util/db/eventDbService';
import { EVENT_TYPES } from '../../../event/event';
import { pollCard } from './pollCard';

const createCardValidationSchema = yup.object({
  ownerId: yup.string()
    .required(),

  billingDetails: billingDetailsValidationSchema
    .required(),

  keyId: yup
    .string()
    .required(),

  sessionId: yup
    .string()
    .required(),

  ipAddress: yup
    .string()
    .required(),

  encryptedData: yup
    .string()
    .required(),

  expMonth: yup
    .number()
    .min(1)
    .max(12)
    .required(),

  expYear: yup
    .number()
    .min(new Date().getFullYear())
    .max(new Date().getFullYear() + 50)
    .required()
});

type CreateCardPayload = yup.InferType<typeof createCardValidationSchema>;

/**
 * Creates card in the FireStore and on Circle side.
 *
 * @param payload
 */
export const createCard = async (payload: CreateCardPayload = {}): Promise<ICardEntity> => {
  // Validate the passed data
  await validate(payload, createCardValidationSchema);

  // Find the creator
  const user = await userDb.get(payload.ownerId);

  // Format the data for circle
  const headers = await getCircleHeaders();
  const data: ICircleCreateCardPayload = {
    billingDetails: payload.billingDetails as any,
    encryptedData: payload.encryptedData,
    expMonth: payload.expMonth,
    expYear: payload.expYear,
    idempotencyKey: payload.sessionId,
    keyId: payload.keyId,
    metadata: {
      email: user.email,
      ipAddress: payload.ipAddress,
      sessionId: payload.sessionId
    }
  };

  // @todo Move this to the circle client
  // Create the card on Circle
  const { data: response } = await externalRequestExecutor<ICircleCreateCardResponse>(async () => {
    logger.debug('Trying to create new card with circle', {
      data
    });

    return (await axios.post<ICircleCreateCardResponse>(`${circlePayApi}/cards`,
      data,
      headers
    )).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    userMessage: 'Cannot create the card, because it was rejected by Circle'
  });

  // Check if the use already has the same card
  const existingCards = await cardDb.getMany({
    ownerId: user.uid,
    circleCardId: response.id
  });

  if (existingCards.length) {
    //  note to self:
    //    this should not end up deployed as we currently
    //    have no UI for that, so we are creating new card
    //    on every request. Just return the found card without
    //    saving it again
    //
    // ========================================================================
    //
    // throw new CommonError(`Cannot resave the same card for the same user!`, {
    //   cardId: response.id,
    //   userId: user.id
    // });
    return existingCards[0];
  }


  // If the card was created successfully save it
  const card = await cardDb.add({
    ownerId: user.uid,
    circleCardId: response.id,
    metadata: {
      billingDetails: data.billingDetails,
      digits: response.last4,
      network: response.network
    },
    verification: {
      cvv: response.verification.cvv as any
    }
  });

  // After the card is saved check the status of the card
  await pollCard(card);

  // Create event
  await createEvent({
    type: EVENT_TYPES.CARD_CREATED,
    userId: user.uid,
    objectId: card.id
  });

  // Return the created card
  return card;
};