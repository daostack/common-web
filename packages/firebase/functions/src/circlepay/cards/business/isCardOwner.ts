import { cardDb } from '../database';

/**
 * Checks if the user is the creator of the card
 *
 * @param userId - The user ID, that we want to verify is the owner
 * @param cardId - The ID of the card, against which we want to check
 */
export const isCardOwner = async (userId: string, cardId: string): Promise<boolean> => {
  return (await cardDb.get(cardId)).ownerId === userId;
};