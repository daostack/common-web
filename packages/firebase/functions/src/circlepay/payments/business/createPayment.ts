import axios from 'axios';
import * as yup from 'yup';
import { v4 } from 'uuid';

import { IPaymentEntity } from '../types';
import { validate } from '../../../util/validate';
import { getCircleHeaders } from '../../index';
import { ICircleCreatePaymentPayload, ICircleCreatePaymentResponse } from '../../types';
import { userDb } from '../../../core/domain/users/database';
import { externalRequestExecutor } from '../../../util';
import { circlePayApi } from '../../../settings';
import { ErrorCodes } from '../../../constants';
import { paymentDb } from '../database';
import { createEvent } from '../../../util/db/eventDbService';
import { EVENT_TYPES } from '../../../event/event';
import { cardDb } from '../../cards/database';
import { CommonError, CvvVerificationError } from '../../../util/errors';

const createPaymentValidationSchema = yup.object({
  userId: yup.string()
    .required(),

  cardId: yup.string()
    .uuid()
    .required(),

  proposalId: yup.string()
    .required(),

  subscriptionId: yup.string()
    .when('type', {
      is: 'subscription',
      then: yup.string().required()
    }),

  ipAddress: yup.string()
    .required(),

  sessionId: yup.string()
    .required(),

  amount: yup.number()
    .required(),

  type: yup.string()
    .oneOf(['one-time', 'subscription'])
    .required(),

  encryptedData: yup.string()
    .when('type', {
      is: 'one-time',
      then: yup.string() //.required()
    }),

  keyId: yup.string()
    .when('type', {
      is: 'one-time',
      then: yup.string() //.required()
    })
});

interface ICreatePaymentPayload extends yup.InferType<typeof createPaymentValidationSchema> {
  /**
   * This is the ID of the object that we are charging (the proposal in this case). It is used
   * as idempotency key, so we don't charge more than one time for one thing
   */
  proposalId: string;

  /**
   * This is the amount that the source will be charged in US dollar cents
   */
  amount: number;
}

/**
 * This function only creates the payment with circle and saves it in the database. No polling is done here.
 * Please do not use directly or use with caution.
 *
 * @throws { CommonError } - If the card is not owned by the user
 *
 * @param payload - The data needed for creating payment
 */
export const createPayment = async (payload: ICreatePaymentPayload): Promise<IPaymentEntity> => {
  // Validate the data
  await validate(payload, createPaymentValidationSchema);

  // Find the payment initiator
  const user = await userDb.get(payload.userId);
  const card = await cardDb.get(payload.cardId);

  // If the user is not the owner of the card throw an error
  if (card.ownerId !== user.uid) {
    throw new CommonError('Cannot charge card, that you do not own', {
      card,
      user
    });
  }

  // If the card has failed CVV verification check do not allow the payment
  // In real cases we should never end up here
  if (card.verification && card.verification.cvv !== 'pass') {
    logger.warn('Payment is attempted on card without CVV verification pass', {
      card,
      paymentCreationPayload: payload
    });

    throw new CvvVerificationError(card.id);
  }

  // Format the data for circle
  const headers = await getCircleHeaders();
  const circleData: ICircleCreatePaymentPayload = {
    amount: {
      // In our code all money values should be in cents
      amount: payload.amount / 100,
      currency: 'USD'
    },

    source: {
      type: 'card',
      id: card.circleCardId
    },

    metadata: {
      email: user.email,
      ipAddress: payload.ipAddress,
      sessionId: payload.sessionId
    },

    idempotencyKey: payload.type === 'one-time'
      ? payload.proposalId
      : v4(), // @todo This will not work for the second subscription payment, so append the
              //    times that the subscription has been charged?

    verification: 'none'
  };

  // Create the request @todo Handle errors better
  const { data: response } = await externalRequestExecutor<ICircleCreatePaymentResponse>(async () => {
    logger.debug('Creating payment in Circle', {
      data: circleData
    });

    return (await axios.post<ICircleCreatePaymentResponse>(`${circlePayApi}/payments`,
      circleData,
      headers
    )).data;
  }, {
    errorCode: ErrorCodes.CirclePayError,
    userMessage: 'Call to CirclePay failed. Please try again later and if the issue persist contact us.'
  });

  // Save the payment
  const payment = await paymentDb.add({
    amount: {
      amount: payload.amount,
      currency: 'USD'
    },

    source: {
      type: 'card',
      id: card.id
    },

    type: payload.type,
    proposalId: payload.proposalId,
    userId: user.uid,
    status: response.status,
    circlePaymentId: response.id,

    // Add the subscription ID to the entity if the payment is for subscription
    ...(payload.subscriptionId && {
      subscriptionId: payload.subscriptionId
    })
  });

  logger.debug('New payment created', {
    payment
  });

  // Create event
  await createEvent({
    type: EVENT_TYPES.PAYMENT_CREATED,
    userId: user.uid,
    objectId: payment.id
  });

  // @todo Schedule polling?

  // Return the created payment
  return payment;
};