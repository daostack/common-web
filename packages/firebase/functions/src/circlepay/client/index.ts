import { getPaymentFromCircle } from './getPaymentFromCircle';
import { getCardFromCircle } from './getCardFromCircle';
import { getBankAccountFromCircle } from './getBankAccountFromCircle';

export { ICircleCard, CircleCardNetwork } from './getCardFromCircle';

export const circleClient = {
  getPayment: getPaymentFromCircle,
  getCard: getCardFromCircle,
  getBankAccount: getBankAccountFromCircle
}