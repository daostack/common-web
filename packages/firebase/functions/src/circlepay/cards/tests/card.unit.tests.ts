import { v4 } from 'uuid';

import '../../../util/tests/helpers/mockers/cards/getCard.mocker';
import '../../../util/tests/helpers/mockers/cards/updateCard.mocker';
import '../../../util/tests/helpers/mockers/firebase.mocker';

import { isCardOwner } from '../business/isCardOwner';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { IGetCircleCardResponse } from '../../client/getCardFromCircle';
import { getCardFailedCvvCheck, getCardPendingChecks, getCardSuccessfulChecks } from './data/testCardCirlceResponses';
import { pollCard } from '../business/pollCard';
import { pendingCardEntity } from './data/testCardEntities';


// Add the logger to the tests globals
import '../../../util/logger';

// ----- Local Mocks

jest.mock('firebase-admin', () => ({
  initializeApp: jest.fn(),
  firestore: () => ({
    collection: jest.fn(() => ({
      withConverter: jest.fn()
        .mockImplementation(() => jest.fn())
      //
      // where: jest.fn(queryString => ({
      //   get: mockQueryResponse
      //
      // }))
    })),
    settings: jest.fn()
  })
}));

// ----- Match helpers

const circleMatcher = /.+circle\.com\/.*/;
const circeCardEndpointMatcher = new RegExp(circleMatcher.source + (/\/cards/).source);

describe('Card unit tests', () => {
  describe('isCardOwner', () => {
    it('should return true if the card user is really the owner', async () => {
      // Arrange
      const ownerId = v4();

      // Act
      // The mocker sets the card ID to be the same as the owner ID
      const result = await isCardOwner(ownerId, ownerId);

      // Assert
      expect(result).toBeTruthy();
    });

    it('should return false if the user is not the owner', async () => {
      // Arrange
      const cardId = v4();
      const userId = v4();

      // Act
      // The mocker sets the card ID to be the same as the owner ID
      const result = await isCardOwner(userId, cardId);

      // Assert
      expect(result).toBeFalsy();
    });
  });

  describe('pollCard', () => {
    it('should finish on card verification success', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');

      mock
        // On the first and second call return pending
        .onGet(circeCardEndpointMatcher)
        .replyOnce<IGetCircleCardResponse>(200, getCardPendingChecks)

        .onGet(circeCardEndpointMatcher)
        .replyOnce<IGetCircleCardResponse>(200, getCardPendingChecks)

        // On the third and all sequential calls return check success
        .onGet(circeCardEndpointMatcher)
        .reply<IGetCircleCardResponse>(200, getCardSuccessfulChecks);

      // Act
      const pollCardResult = await pollCard(pendingCardEntity);

      // Assert
      delete pollCardResult.updatedAt;
      delete pollCardResult.createdAt;

      expect(pollCardResult).toMatchSnapshot();
      expect(pollCardResult.verification.cvv).toBe('pass');

      expect(mockSpy).toHaveBeenCalledTimes(3);

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });

    it('should finish on card verification failure', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');

      mock
        // On the first and second call return pending
        .onGet(circeCardEndpointMatcher)
        .replyOnce<IGetCircleCardResponse>(200, getCardPendingChecks)

        .onGet(circeCardEndpointMatcher)
        .replyOnce<IGetCircleCardResponse>(200, getCardPendingChecks)

        // On the third and all sequential calls return check success
        .onGet(circeCardEndpointMatcher)
        .reply<IGetCircleCardResponse>(200, getCardFailedCvvCheck);

      // Act
      const pollCardResult = await pollCard(pendingCardEntity, {
        throwOnCvvFail: false
      });

      // Assert
      expect(pollCardResult).toBeTruthy();
      expect(pollCardResult.ownerId).toBeTruthy();
      expect(pollCardResult.verification.cvv).toBe('fail');

      expect(mockSpy).toHaveBeenCalledTimes(3);

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });

    it('should throw on card verification failure if that is requests', async () => {
      // Arrange
      const mock = new MockAdapter(axios);
      const mockSpy = jest.spyOn(axios, 'get');

      mock
        // On the first and second call return pending
        .onGet(circeCardEndpointMatcher)
        .replyOnce<IGetCircleCardResponse>(200, getCardPendingChecks)

        // On the second and all sequential calls return check success
        .onGet(circeCardEndpointMatcher)
        .reply<IGetCircleCardResponse>(200, getCardFailedCvvCheck);

      // Act & Assert
      await expect(pollCard(pendingCardEntity, { throwOnCvvFail: true }))
        .rejects
        .toThrowError('CVV verification failed for card');

      expect(mockSpy).toHaveBeenCalledTimes(2);

      // Cleanup
      mock.restore();
      mockSpy.mockClear();
    });
  });
});