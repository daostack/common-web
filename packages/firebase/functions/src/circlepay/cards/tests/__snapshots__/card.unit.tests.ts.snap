// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card unit tests pollCard should finish on card verification success 1`] = `
Object {
  "circleCardId": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa75",
  "id": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa76",
  "metadata": Object {
    "billingDetails": Object {
      "city": "Sofia",
      "country": "Bulgaria",
      "line1": "Nqkade v Sofia",
      "line2": "",
      "name": "Alex Ivanov",
      "postalCode": "1000",
    },
    "digits": "0832",
    "network": "VISA",
  },
  "ownerId": "ae6b689c-5dcb-41ed-84a7-51e6ce7faa72",
  "verification": Object {
    "cvv": "pass",
  },
}
`;
